File: frontend/netlify/functions/encounters.py
"""
Netlify Serverless Function: Encounter Management
Endpoint: /.netlify/functions/encounters
Methods: GET, POST, PUT, DELETE
"""

import json
import logging
import sys
import os
import re
from datetime import datetime

# Add shared modules to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'shared'))

from db_utils import (
    get_db_connection,
    put_db_connection,
    create_encounter,
    fetch_encounters_by_patient,
    fetch_encounter_by_id,
    update_encounter,
    delete_encounter,
    DatabaseError,
    EncounterNotFoundError
)
from pydantic import BaseModel, Field, ValidationError
from typing import Optional
from datetime import date, time

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class EncounterCreateData(BaseModel):
    """Pydantic model for encounter creation"""
    patient_id: int = Field(..., gt=0)
    encounter_date: Optional[date] = None
    encounter_time: Optional[time] = None
    chief_complaint: Optional[str] = None
    notes: Optional[str] = None
    user_id: Optional[int] = None


def handler(event, context):
    """Netlify serverless function handler for encounter management."""
    headers = {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
    }

    if event.get('httpMethod') == 'OPTIONS':
        return {'statusCode': 200, 'headers': headers, 'body': ''}

    conn = None
    try:
        conn = get_db_connection()
        if not conn:
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps({'error': 'Database connection failed'})
            }

        method = event.get('httpMethod')
        path = event.get('path', '')

        # Extract IDs from path
        encounter_match = re.search(r'/encounters/(\d+)', path)
        patient_match = re.search(r'/patients/(\d+)/encounters', path)

        encounter_id = int(encounter_match.group(1)) if encounter_match else None
        patient_id = int(patient_match.group(1)) if patient_match else None

        if method == 'GET':
            if encounter_id:
                return get_encounter(conn, encounter_id, headers)
            elif patient_id:
                return get_patient_encounters(conn, patient_id, headers)
            else:
                return {'statusCode': 400, 'headers': headers,
                        'body': json.dumps({'error': 'Missing patient_id or encounter_id'})}

        elif method == 'POST' and patient_id:
            return create_new_encounter(conn, patient_id, event, headers)

        elif method == 'PUT' and encounter_id:
            return update_existing_encounter(conn, encounter_id, event, headers)

        elif method == 'DELETE' and encounter_id:
            return delete_existing_encounter(conn, encounter_id, headers)

        else:
            return {'statusCode': 400, 'headers': headers, 'body': json.dumps({'error': 'Invalid request'})}

    except Exception as e:
        logger.exception(f"Error in encounter management: {e}")
        if conn:
            conn.rollback()
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'error': 'Internal server error'})
        }
    finally:
        if conn:
            put_db_connection(conn)


def get_patient_encounters(conn, patient_id, headers):
    """Get all encounters for a patient"""
    try:
        encounters = fetch_encounters_by_patient(conn=conn, patient_id=patient_id)

        # Serialize time objects
        for encounter in encounters:
            if isinstance(encounter.get('encounter_time'), time):
                encounter['encounter_time'] = encounter['encounter_time'].strftime("%H:%M")

        return {'statusCode': 200 if encounters else 404, 'headers': headers,
                'body': json.dumps(encounters if encounters else {'error': 'No encounters found'})}
    except Exception as e:
        logger.error(f"Error: {e}")
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}


def get_encounter(conn, encounter_id, headers):
    """Get a single encounter"""
    try:
        encounter = fetch_encounter_by_id(conn=conn, encounter_id=encounter_id)
        if encounter:
            if 'encounter_time' in encounter and isinstance(encounter['encounter_time'], time):
                encounter['encounter_time'] = encounter['encounter_time'].isoformat()
            return {'statusCode': 200, 'headers': headers, 'body': json.dumps(encounter)}
        return {'statusCode': 404, 'headers': headers, 'body': json.dumps({'error': 'Encounter not found'})}
    except Exception as e:
        logger.error(f"Error: {e}")
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}


def create_new_encounter(conn, patient_id, event, headers):
    """Create a new encounter"""
    try:
        body = json.loads(event['body']) if isinstance(event.get('body'), str) else event.get('body', {})

        # Auto-set date and time
        now = datetime.now()
        body["encounter_date"] = now.date().isoformat()
        body["encounter_time"] = now.time().strftime("%H:%M:%S")

        validated_data = EncounterCreateData.model_validate({**body, "patient_id": patient_id})

        encounter_id = create_encounter(
            conn=conn,
            patient_id=validated_data.patient_id,
            encounter_date=validated_data.encounter_date,
            encounter_time=validated_data.encounter_time,
            chief_complaint=validated_data.chief_complaint,
            notes=validated_data.notes,
            user_id=validated_data.user_id
        )

        if encounter_id:
            new_encounter = fetch_encounter_by_id(conn=conn, encounter_id=encounter_id)
            return {'statusCode': 201, 'headers': headers,
                    'body': json.dumps({'message': 'Encounter created', 'encounter': new_encounter})}
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': 'Failed to create'})}
    except ValidationError as e:
        return {'statusCode': 400, 'headers': headers, 'body': json.dumps({'error': e.errors()})}
    except Exception as e:
        logger.error(f"Error: {e}")
        conn.rollback()
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}


def update_existing_encounter(conn, encounter_id, event, headers):
    """Update an encounter"""
    try:
        body = json.loads(event['body']) if isinstance(event.get('body'), str) else event.get('body', {})
        updated_rows = update_encounter(
            conn=conn,
            encounter_id=encounter_id,
            encounter_date=body.get('encounter_date'),
            encounter_time=body.get('encounter_time'),
            chief_complaint=body.get('chief_complaint'),
            notes=body.get('notes'),
            user_id=body.get('user_id')
        )

        if updated_rows > 0:
            updated = fetch_encounter_by_id(conn=conn, encounter_id=encounter_id)
            return {'statusCode': 200, 'headers': headers,
                    'body': json.dumps({'message': 'Updated successfully', 'encounter': updated})}
        return {'statusCode': 404, 'headers': headers, 'body': json.dumps({'error': 'Not found'})}
    except Exception as e:
        logger.error(f"Error: {e}")
        conn.rollback()
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}


def delete_existing_encounter(conn, encounter_id, headers):
    """Delete an encounter"""
    try:
        deleted = delete_encounter(conn=conn, encounter_id=encounter_id)
        if deleted > 0:
            return {'statusCode': 200, 'headers': headers, 'body': json.dumps({'message': 'Deleted successfully'})}
        return {'statusCode': 404, 'headers': headers, 'body': json.dumps({'error': 'Not found'})}
    except Exception as e:
        logger.error(f"Error: {e}")
        conn.rollback()
        return {'statusCode': 500, 'headers': headers, 'body': json.dumps({'error': str(e)})}

File: frontend/netlify/functions/encounters.js
const { Pool } = require('pg');

const pool = new Pool({
  connectionString: process.env.DATABASE_URL,
  ssl: process.env.NODE_ENV === 'production' ? { rejectUnauthorized: false } : false
});

const headers = {
  'Content-Type': 'application/json',
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type'
};

exports.handler = async (event) => {
  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 200, headers, body: '' };
  }

  const encounterMatch = event.path.match(/\/encounters\/(\d+)/);
  const patientMatch = event.path.match(/\/patients\/(\d+)\/encounters/);

  const encounterId = encounterMatch ? parseInt(encounterMatch[1]) : null;
  const patientId = patientMatch ? parseInt(patientMatch[1]) : null;

  try {
    switch (event.httpMethod) {
      case 'GET':
        if (encounterId) return await getEncounter(encounterId);
        if (patientId) return await getPatientEncounters(patientId);
        return { statusCode: 400, headers, body: JSON.stringify({ error: 'Patient ID or Encounter ID required' }) };
      case 'POST':
        if (!patientId) return { statusCode: 400, headers, body: JSON.stringify({ error: 'Patient ID required' }) };
        return await createEncounter(patientId, JSON.parse(event.body || '{}'));
      case 'PUT':
        if (!encounterId) return { statusCode: 400, headers, body: JSON.stringify({ error: 'Encounter ID required' }) };
        return await updateEncounter(encounterId, JSON.parse(event.body || '{}'));
      case 'DELETE':
        if (!encounterId) return { statusCode: 400, headers, body: JSON.stringify({ error: 'Encounter ID required' }) };
        return await deleteEncounter(encounterId);
      default:
        return { statusCode: 405, headers, body: JSON.stringify({ error: 'Method not allowed' }) };
    }
  } catch (error) {
    console.error('Encounter operation error:', error);
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ error: 'Internal server error', details: error.message })
    };
  }
};

async function getPatientEncounters(patientId) {
  const result = await pool.query(
    'SELECT * FROM encounters WHERE patient_id = $1 ORDER BY encounter_date DESC',
    [patientId]
  );
  return { statusCode: 200, headers, body: JSON.stringify(result.rows) };
}

async function getEncounter(id) {
  const result = await pool.query('SELECT * FROM encounters WHERE encounter_id = $1', [id]);
  if (result.rows.length === 0) {
    return { statusCode: 404, headers, body: JSON.stringify({ error: 'Encounter not found' }) };
  }
  return { statusCode: 200, headers, body: JSON.stringify(result.rows[0]) };
}

async function createEncounter(patientId, data) {
  const result = await pool.query(
    `INSERT INTO encounters (patient_id, encounter_date, encounter_time, chief_complaint, notes, user_id)
     VALUES ($1, $2, $3, $4, $5, $6) RETURNING *`,
    [patientId, data.encounter_date, data.encounter_time, data.chief_complaint, data.notes, data.user_id]
  );
  return {
    statusCode: 201,
    headers,
    body: JSON.stringify({ message: 'Encounter created successfully', encounter: result.rows[0] })
  };
}

async function updateEncounter(id, data) {
  const fields = [];
  const values = [];
  let idx = 1;

  if (data.encounter_date) { fields.push(`encounter_date = $${idx++}`); values.push(data.encounter_date); }
  if (data.encounter_time) { fields.push(`encounter_time = $${idx++}`); values.push(data.encounter_time); }
  if (data.chief_complaint) { fields.push(`chief_complaint = $${idx++}`); values.push(data.chief_complaint); }
  if (data.notes !== undefined) { fields.push(`notes = $${idx++}`); values.push(data.notes); }

  if (fields.length === 0) {
    return { statusCode: 400, headers, body: JSON.stringify({ error: 'No fields to update' }) };
  }

  fields.push(`updated_at = NOW()`);
  values.push(id);

  const result = await pool.query(
    `UPDATE encounters SET ${fields.join(', ')} WHERE encounter_id = $${idx} RETURNING *`,
    values
  );

  if (result.rows.length === 0) {
    return { statusCode: 404, headers, body: JSON.stringify({ error: 'Encounter not found' }) };
  }
  return {
    statusCode: 200,
    headers,
    body: JSON.stringify({ message: 'Encounter updated successfully', encounter: result.rows[0] })
  };
}

async function deleteEncounter(id) {
  const result = await pool.query('DELETE FROM encounters WHERE encounter_id = $1 RETURNING *', [id]);
  if (result.rows.length === 0) {
    return { statusCode: 404, headers, body: JSON.stringify({ error: 'Encounter not found' }) };
  }
  return { statusCode: 200, headers, body: JSON.stringify({ message: 'Encounter deleted successfully' }) };
}
